### PR TITLE
Create Undo & Redo Command 

### LINKED ISSUE #89: Fix Add Bugs
1. Extra Fields Error does not behave as expected

Command to replicate:
```
/pooch-add ; name : Pooch ; extra-field : extra field ; phone : 98883888 ; address : Pooch Street 32 ; email : impooch@gmail.com
```

Current Behaviour:
`extra-field` is parsed part of the previous prefix. In the above example, `name` will be read as 'Pooch ; extra-field : extra field'.

Expected Behaviour:
Error message "Pooch Staff/Supplier/Maintenance does not contain a [additional field] field. 🐾" thrown.

2. Missing Fields Error does not throw the right error

Command to replicate:
```
/pooch-add ; name : Pooch ; address : Pooch Street 32 ; email : impooch@gmail.com
```

Current Behaviour:
Error message "Invalid command format!" is thrown.

Expected Behaviour:
Error message "Pooch Staff/Supplier/Maintenance requires a name/phone/address/email/salary/commission/employment field. 🐾" thrown.

### LINKED ISSUE #95: Fix ArgumentTokeniser and ArgumentMultimap truncation bug
There is a major issue with how ArgumentTokeniser and ArgumentMultimap deals w prefixes which affects Add, Edit and Search commands. It does not truncate the ending string if a field entered is not a valid prefix.

Typing in this sequence of code:
```
/pooch-add ; name : poochies ; phone : 98765432 ; email : a@a.com ; address : 3 ; eeeee : aaaa
/edit ; name : poochies ; field : {address : aaa ; eee : eee}
/search ; name : poochies
/search ; name : poochies ; eee : aaa
```

These are the following buggy outputs:

`/pooch-add ; name : poochies ; phone : 98765432 ; email : a@a.com ; address : 3 ; eeeee : aaaa`
image

`/edit ; name : poochies ; field : {address : aaa ; eee : eee}`
image

`/search ; name : poochies`
image
1 person listed which is poochie.

`/search ; name : poochies ; eee : aaa`
image
0 persons listed because the tokeniser detects the search entry as "poochies ; eee : aaa"

### LINKED ISSUE #97: Add undo/redo command

### LINKED ISSUE #110: Update DG (add, undo, redo command)

### LINKED ISSUE #111: Update UG (add, undo, redo command)

### PR DESCRIPTION
**Pull request by chiageng**
Fixes #89 
Fixes #95 
Fixes #97 
Fixes #110 
Fixes #111 

### New function in ArgumentTokenizer : checkUnknownPrefix(String, Prefix...)
- This function will **return the unknown prefix as string**
- Updated on **catching unknown prefix in Add Commands**
- Developers can repeatedly **use this function to catch unknown prefix** on each of the Command.

### New Command : Undo and Redo
- **UndoCommand and RedoCommand** classes are created to support undo and redo operation. 

### Changes on Address Book
- Original Address Book doesnt store any version.
- Created **new class VersionedAddressBook extends from AddressBook class**
- VersionedAddressBook will **keep a new state into History Tracker** whenever addCommand, editCommand, deleteCommand are called. 
- VersionedAddressBook will keep track of state in an array, **by shifting the pointer on the History Tracker** which will allow undo / redo command to be executed.

### Changes on ModelManager
- ModelManager will **contain VersionedAddressBook instead of AddressBook**
- New operations supported by ModelManager : **commitAddressBook(), undoAddressBook, redoAddressBook(), canUndo(), canRedo().**
- Whenever **modification is done on address book** (such as add / edit / delete), ModelManager will **execute** **commitAddressBook()** in order to **store new state into VersionedAddressBook.**
- When **at least one commit** is done on addressbook, **undo command only able to execute**. The validity of the execution will be checked by using canUndo().
- When **at least one undo** is done on addressbook, **redo command only able to execute**. The validify of the execution will be checked by using canRedo().
- Whenever **new commit is done** on addressbook, the **"future state" of address book arised due to undo command will all be refreshed and no longer able to execute redo**. To execute redo, undo command has to be executed at least once from this state.

## Requires Extra Attention from Developers
- During testing, be aware of followings point when using ModelManager and VersionedAddressBook.
- After **any execution** of command, **ModelManager will do** a **commit on VersionedAddressBook**, it will have **different size of History state** as before, and so it is **considered as different VersionedAddressBook** compared to before the execution. 
- Whenever **undo / redo command is executed**, the pointer in VersionedAddressBook is **pointing at different state**, hence it is **considered as different VersionedAddressBook** compared to before the execution.

### CODE REVIEW

File: src/main/java/seedu/address/model/VersionedAddressBook.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Updates the sublist of address book states up to the current state pointer.
+     */
+    private void updateSubAddressBookStateList() {
+        ArrayList<ReadOnlyAddressBook> copy = new ArrayList<>();
```
Comment (yleeyilin) [inline src/main/java/seedu/address/model/VersionedAddressBook.java:77]: I think the naming of this variable can be better. I.e newAddressBookList or something.

**Review by yleeyilin — APPROVED** [review]
LGTM!

**Review by jannaleong — APPROVED** [review]
LGTM!

**Review by Joshy837 — APPROVED** [review]
LGTM

**Review by jamessinmaojun — APPROVED** [review]
lgtm

**Pull request merged**